app/Http/Controllers/V4DB/MangaController.php

```diff
@@ -730,9 +730,19 @@ public function reviews(Request $request, int $id)
             $results->isEmpty()
             || $this->isExpired($request, $results)
         ) {
-            $page = $request->get('page') ?? 1;
-            $manga = $this->jikan->getMangaReviews(new MangaReviewsRequest($id, $page));
-            $response = \json_decode($this->serializer->serialize($manga, 'json'), true);
+
+            $response = ScraperHelper::getSerializedJSON(
+                app('JikanParser')
```

Comment:
Service locator pattern is an anti-pattern. Laravel allows it, but the best way to do this would be to use the ctor of the `ScraperHelper`, register `ScraperHelper` as a service, and add it to the ctor  of the controller.
I know that you would argue this form:
`$this->scraperHelper->xyz()`   
But that's the nice way to do it afaik: https://softwareengineering.stackexchange.com/questions/179002/ioc-containers-and-service-locator-pattern

Reply:
Just to show you what I'm talking about:

```php
// this is in app/Providers
class AppServiceProvider extends ServiceProvider
{
    public function register(): void
    {
        $this->app->singleton(Serializer::class, function ($app) {
            return SerializerFactory::createV4();
        });
    }
}

// forgot the syntax for interfaces in php, this is just a rough example
// we use the interface as a "service contract"
interface IScraperHelper
{
   getSerializedJSON(object|array $response): array;
}

// this is the default service implementation of service contract "IScraperHelper"
class DefaultScraperHelper implements IScraperHelper
{
    private Serializer $serializer;

    // laravel will automatically inject this
    // this is the expression of the dependency
    public function __contruct(Serializer $serializer)
    {
        $this->serializer = $serializer;
    }

    public function getSerializedJSON(object|array $response): array
    {
        return \json_decode($this->serializer->serialize($response, 'json'), true);
    } 
}

class MyController extends Controller
{
    private IScarperHelper $scraperHelper;
    private JikanParser $parser;

    public function __contruct(IScraperHelper $scaperHelper, JikanParser $parser)
    {
         $this->scraperHelper = $scraperHelper;
         $this->parser = $parser;
    }

    public function index()
    {
        // even this could be wrapped in a service
        return $this->scraperHandler->getSerializedJSON($this->parser->getMangaReviews(/* ... */));
    }
}
```

I was talking about this kind of style of coding. ^
The aim with this is to have smaller testable units of code. Of course you can get this one wrong too, and create too many of them. 😄

Reply:
Another example (not the best one) from my days in university, a card game in unity, but using the same pattern: https://github.com/pushrbx/unitysnapgame/tree/master/Logic/SnapGameLogic/Abstractions

Reply:
@pushrbx Thanks for the elaborate response, I will look into this. Let's refactor this in a separate PR (we can target other endpoints as well) and merge this one for a new release tag for now.